appliance/mariadb/process.go

```diff
@@ -0,0 +1,1106 @@
+				logger.Error("error reading master status", "err", err)
+				startTime = time.Now().UTC()
+				select {
+				case <-stopCh:
+					logger.Debug("canceled, stopping")
+					return
+				case <-time.After(checkInterval):
+				}
+				continue
+			}
+
+			// Read downstream slave status.
+			ss, err := p.slaveStatus(downstream.Addr)
+			if err != nil {
+				logger.Error("error reading slave status", "err", err)
+				startTime = time.Now().UTC()
+				select {
+				case <-stopCh:
+					logger.Debug("canceled, stopping")
+					return
+				case <-time.After(checkInterval):
+				}
+				continue
+			}
+
+			elapsedTime := time.Since(startTime)
+			logger := logger.New(
+				"master_log_file", ms.File, "master_log_pos", ms.Position,
+				"slave_log_file", ss.MasterLogFile, "slave_log_pos", ss.ReadMasterLogPos,
+				"elapsed", elapsedTime,
+			)
+
+			// Update the slave status and reset start time if replication has progressed.
+			if cmp := ComparePosition(status.MasterLogFile, status.ReadMasterLogPos, ss.MasterLogFile, ss.ReadMasterLogPos); status.IsZero() || cmp == -1 {
+				logger.Debug("slave status progressing, resetting start time")
+				startTime = time.Now().UTC()
+				status = ss
+			}
+
+			logger.Info("downstream status", "master_file", ms.File, "master_pos", ms.Position, "slave_file", ss.MasterLogFile, "slave_pos", ss.ReadMasterLogPos)
+			if ms.File == ss.MasterLogFile && ms.Position == ss.ReadMasterLogPos {
+				logger.Info("downstream caught up")
+				p.syncedDownstreamValue.Store(downstream)
+				break
+			}
+
+			if elapsedTime > p.ReplTimeout {
+				logger.Error("error checking replication status", "err", "downstream unable to make forward progress")
+				return
+			}
+
+			logger.Debug("continuing replication check")
+			select {
+			case <-stopCh:
+				logger.Debug("canceled, stopping")
+				return
+			case <-time.After(checkInterval):
+			}
+		}
+
+		if enableWrites {
+			// TODO: enable writes, restart?
+		}
+	}()
+}
+
+// DSN returns the datasource name for a host.
+func (p *Process) DSN(host, database string) *DSN {
+	return &DSN{
+		Host:     host,
+		Database: database,
+		User:     "flynn",
+		Password: p.Password,
+		Timeout:  p.OpTimeout,
+	}
+}
+
+func (p *Process) masterStatus(host string) (status MasterStatus, err error) {
+	logger := p.Logger.New("fn", "masterStatus", "host", host)
+
+	db, err := sql.Open("mysql", p.DSN(host, "").String())
+	if err != nil {
+		return
+	}
+	defer db.Close()
+
+	var discard interface{}
+	if err = db.QueryRow("SHOW MASTER STATUS").Scan(
+		&status.File,
+		&status.Position,
+		&discard, // Binlog_Do_DB
+		&discard, // Binlog_Ignore_DB
+	); err == sql.ErrNoRows {
+		err = ErrNoReplicationStatus
+		return
+	} else if err != nil {
+		return
+	}
+
+	logger.Info("master status", "file", status.File, "position", status.Position)
+
+	return
+}
+
+// XLogPosition returns the current transaction log position.
+func (p *Process) XLogPosition() (xlog.Position, error) { panic("FIXME") }
```

Comment:
I did a bit more reading. I think we should look into the MySQL GTID (Global Transaction ID). It seems to be a monotonically increasing counter that is persisted through the entire replication hierarchy. Seems like a much better solution than the file/offset setup which seems like it isn't comparable between layers of replication.